Import dive sites from XML into dive site table

djunkins opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
Adds feature to import dive sites from an XML file (or multiple files) into the dive site table for a user. A dialog box is provided to allow the user to select which dive sites to import with sites that are in the same location as existing sites automatically deselected. The nearest existing site and distance to that site are shown in the dialog.
<!-- Describe your pull request in detail. -->

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->
1) Menu option for "Import dive sites" is added under the Import menu
2) Files are parsed to add imported sites to a dive site table
2) File selection window appears to selected the XML files to import sites from
3) Dialog box is added to select the sites to import.
4) Selected sites are added to the core dive_site_table with undo/redo commands

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
CHANGELOG.md has been updated.

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@bstoeger

dirkhh commented:
@bstoeger I'd really appreciate a review from you.
I'd also love to refactor the commits to not have something named "Dummy commit for Doug" in our commit history - but I'll be happy to do that part once @bstoeger approves the code

bstoeger commented:
It's on my list, but please give me a few days. I'm currently travelling, therefore very short on time.

djunkins commented:
Sorry -- I can refactor the commits if you want. I wasn't sure if it was ok to change @bstoeger's commit message since he signed it.

bstoeger commented:
Feel free to change it. I can also reword it, if you prefer.

djunkins commented:
OK. I'll refactor the commits into something more coherent once the code review is complete and we're all in agreement that I'm on the right track. No hurry -- I've been traveling too which is why this took so long to complete.

djunkins commented:
Pushed an updated commit to fix a corner case of trying to compute distance to a nonexistent site when you try to import a site and no sites exist yet. Also rebased against the current master to pull in the get_distance fix.

djunkins commented:
@bstoeger Thanks for the feedback. I was out today, but will work on fixing the issues you raised over this weekend. Some of them may take me some time to research.

djunkins commented on desktop-widgets/divesiteimportdialog.h:
```diff
@@ -0,0 +1,40 @@
+class DivesiteImportDialog : public QDialog {
+	Q_OBJECT
+public:
+	explicit DivesiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported = NULL, QString source = "");
```
@bstoeger I want to be sure I understand what you are saying in your second comment above.

I believe what you are saying is that since `DivesiteImportDialog` is being passed a pointer to a `struct dive_site_table`, it should not be the responsibility of `DivesiteImportDialog` to free the memory used by the table -- the code that called (or created) the `DivesiteImportDialog` object should be responsible for freeing the memory.

Regarding the second part of your comment about the table being cleared in the destructor, I'm not sure what you are asking.

bstoeger replied:
>  believe what you are saying is that since DivesiteImportDialog is being passed a pointer to a struct dive_site_table, it should not be the responsibility of DivesiteImportDialog to free the memory used by the table -- the code that called (or created) the DivesiteImportDialog object should be responsible for freeing the memory.

Actually I'm not saying what it should do, I'm just asking to comment what it does. E.g.
```
// Caller keeps ownership of "imported". The contents of "imported" will be consumed on execution of the dialog.
// On return, it will be empty.
```

The problem is that ownership semantics are very subtle here. What I think it *should* do, but that is not a requirement for this commit:
The class should use its own table as a subobject:
```
class DivesiteImportDialog {
private:
	struct dive_site_table imported_sites;
	...
};
```
and take ownership of the contents in the constructor:
```
// Consumes all elements of "imported". Table will be empty on return of the constructor
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported, QString source)
{
	importedSites = imported;
	imported.nr = imported.allocated = 0;
	imported.dive_sites = nullptr;
}
```

And then in the destructor clean up:
```
DiveSiteImportDialog::~DiveSiteImportDialog()
{
	clear_dive_site_table(&importedSites);
}
```
That's a poor man's implementation of "move semantics".
But that would of course mean changing all `importedSites->` to `importedSites.`, etc. So really not needed for now.

Just another comment: By convention, the `QObject *parent` should always be the last argument.

djunkins replied:
OK. I think I understand how you are saying you would do it, but I want to make sure of one thing. In your third code block above, in the definition of the constructor, I believe it should be `struct dive_site_table imported` instead of `struct dive_site_table *imported` since you are then assigning it to the `struct dive_site_table` subobject you declared in the header file in the second code block.

I also wanted to verify in your examples that `imported_sites` in the second code block should be the same name as `importedSites` in the last two code blocks. Is `imported_sites` the preferred name since it is a C structure instead of a C++ object?

bstoeger replied:
> In your third code block above, in the definition of the constructor, I believe it should be `struct dive_site_table imported` instead of `struct dive_site_table *imported` since you are then assigning it to the `struct dive_site_table` subobject you declared in the header file in the second code block.

Yes, two possibilities:
```
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table *imported, QString source)
{
	importedSites = *imported;
```
or
```
DiveSiteImportDialog::DiveSiteImportDialog(QWidget *parent = 0, struct dive_site_table &imported, QString source)
{
	importedSites = imported;
```

`struct dive_site_table imported` as argument type would be wrong, because it being a "dumb" C-object, the compiler would just copy the innards and now you have two tables using the same allocated memory -> crash by double-free().

But I wouldn't mind if you keep it as it is for now.

> Is imported_sites the preferred name since it is a C structure instead of a C++ object?

I think generally we use camelCase when in C++ code. But we're not very consistent about it. Better ask @dirkhh. :)

bstoeger commented:
> Some of them may take me some time to research.

Note that these are all either coding-style issues or what I believe to be unused code, so it should all be rather simple. Just let me know if my comments were too brief.

djunkins commented on desktop-widgets/mainwindow.cpp:
```diff
@@ -1748,6 +1761,35 @@ void MainWindow::on_actionImportDiveLog_triggered()
 	}
 }
 
+void MainWindow::on_actionImportDiveSites_triggered()
```
I will take another look at it, but I do want to use as much of the existing XML parsing as possible so that it remains consistent. The dive site export uses the same XML generation as the dive log export. I do think it is useful to be able to take a full dive log file and pull the sites out of it for import, but there may be a more efficient way to do it than your test function.

bstoeger replied:
> I do think it is useful to be able to take a full dive log file and pull the sites out of it for import

Fair enough. Speeding up the XML-parser is on my TODO list anyway. :)

dirkhh replied:
Actually, I think this is super-clever, even if this may not have been initially intended.
This way I can skip the whole "export all my sites" thing if I want to sync dive sites between two divers (let's say two buddies). I just import diver 1's XML divelog in dive site import mode and get all of their sites, but none of their dives.
Or am I missing something?

djunkins replied:
@dirkhh I agree that this would allow another way to share all dive sites, but I think there will be times when a user may want to share all of their dive sites, but not all of the dive log and trip data. 

Ultimately what I would like to do is to remove the "dive sites" option from the export dialog altogether and add select boxes to the dive sites tab so that the user can select one, several, or all dive sites and hit an "Export selected dive sites" button. We could also move the "Import dive sites" trigger from the Import menu to a button on the dive sites tab to keep all of the dive site management functions in one place. But before we do that, I'd like to start getting some feedback from other users on the dive site import/export functions as they stand now.

djunkins commented on qt-models/divesiteimportmodel.cpp:
```diff
@@ -0,0 +1,178 @@
+	if (role == Qt::DisplayRole) {
+		switch (column) {
+		case 0:
+			return QVariant(ds->name);
```
I thought the model needed to return a QT object instead of a string.

bstoeger replied:
Yes it should. But you don't have to tell it explicitly to convert to QVariant. `return ds->name' should do that implicitly.

djunkins commented on qt-models/divesiteimportmodel.cpp:
```diff
@@ -0,0 +1,178 @@
+void DivesiteImportedModel::changeSelected(QModelIndex clickedIndex)
+{
+	checkStates[clickedIndex.row()] = !checkStates[clickedIndex.row()];
+	dataChanged(index(clickedIndex.row(), 0), index(clickedIndex.row(), 0), QVector<int>() << Qt::CheckStateRole << Selected);
```
Do you mean that `QVector<int>() << Qt::CheckStateRole << Selected` could be replaced with `QVector<int> { Qt::CheckStateRole, Selected }`?

bstoeger replied:
Yes, the latter looks more "natural" to me. But we have more of the former. Up to you.

djunkins commented on qt-models/divesiteimportmodel.h:
```diff
@@ -0,0 +1,39 @@
+	Qt::ItemFlags flags(const QModelIndex &index) const;
+	Q_INVOKABLE void clearTable();
+	QHash<int, QByteArray> roleNames() const;
+	Q_INVOKABLE void repopulate(dive_site_table_t *sites);
```
I removed the Q_INVOKABLE since it doesn't appear to make sense to support QML for this.

djunkins commented on qt-models/divesiteimportmodel.cpp:
```diff
@@ -0,0 +1,178 @@
+	endResetModel();
+}
+
+QHash<int, QByteArray> DivesiteImportedModel::roleNames() const {
```
It probably does not make sense to support QML for importing dive sites.

bstoeger replied:
I believe that the whole roleNames shenanigans is not necessary if we don't support QML. The data function itself does not change.

djunkins commented on qt-models/divesiteimportmodel.cpp:
```diff
@@ -0,0 +1,178 @@
+	return QAbstractTableModel::flags(index) | Qt::ItemIsUserCheckable;
+}
+
+void DivesiteImportedModel::clearTable()
```
I removed it.

djunkins commented on qt-models/divesiteimportmodel.h:
```diff
@@ -0,0 +1,39 @@
+	int rowCount(const QModelIndex& index = QModelIndex()) const;
+	QVariant data(const QModelIndex& index, int role) const;
+	QVariant headerData(int section, Qt::Orientation orientation, int role) const;
+	void setImportedDivesitesIndexes(int first, int last);
```
I removed the declaration.

djunkins commented on desktop-widgets/divesiteimportdialog.cpp:
```diff
@@ -0,0 +1,79 @@
+	ui->selectAllButton->setEnabled(true);
+	ui->unselectAllButton->setEnabled(true);
+
+	connect(ui->importedDivesitesView, SIGNAL(clicked(QModelIndex)), divesiteImportedModel, SLOT(changeSelected(QModelIndex)));
```
I changed the three connections that were connected to UI objects, but was not sure if it is possible to change the close or quit connections. Looking through the other files in desktop-widgets, I was not able to find an example of how to to this.

djunkins commented:
Three changes in the last push:

1) I removed all of the roleNames cruft from the divesiteimportmodel.cpp that was there to support QML
2) I fixed the type declaration in the DivesiteImportDialog constructor to add the '&' to the imported dive sites variable
3) Made the variable names more consistently camelCase in the C++ files

dirkhh commented:
This looks good to me, but I'd rather see @bstoeger approve it :-)

bstoeger commented:
@djunkins: Have you tested the new version?

Then from my side it's good to go. Personally, I would still remove the redundant `QVariant(...)` returns, but no biggie.

djunkins commented:
@bstoeger I have tested the new version and it seems to work as before. I missed removing the redundant QVariant() returns. Let me do that, test one more time and push the (hopefully) final version.

djunkins commented:
Uggghhhh. I thought I had committed my changes before I pushed, but apparently I missed adding the right file. Trying again...

djunkins commented:
Travis CI build failed on the TestGitStorage::testGitStorageCloudMerge() test -- I'm not sure what I did between the last push and this one to deserve that. :-) 

Any pointers on how to start to figure out what has happened?

atdotde commented:
You can run the tests locally to figure that out.

bstoeger commented:
Apparently a hick-up. This can happen when two instances access the cloud concurrently. Let's commit before it breaks again. :)

djunkins commented:
Thanks @bstoeger. I think that this should allow us to close issue #2041 as well. It was a feature request for sharing dive locations between users that came in while we were working on the import and export of dive sites.

dirkhh commented:
Pro tip: when you add a line

`Fixes #2041 `

to your commit message, then this automagically closes the issue when merged :-)
